packages/server/src/adapters/node-http/nodeHTTPRequestHandler.ts

```diff
@@ -32,6 +32,23 @@ export async function nodeHTTPRequestHandler<
   };
   const { path, router } = opts;
 
+  if (opts.cors) {
+    await new Promise<void>((resolve, reject) => {
+      opts.cors!(opts.req, opts.res, (err) => {
+        if (err) {
+          reject(err);
+        } else {
+          resolve();
+        }
+      });
+    });
+
+    if (opts.res.writableEnded) {
+      // cors handled the request
+      return;
+    }
```

Comment:
This is a little unconventional, but the alternative is to wrap the rest of the function in the `next` callback, which is fine, but a bigger refactor

Reply:
Does this even need to be a promise?

Reply:
I feel like a `connect`-style-middleware would be good

Reply:
Nope, only used here to unwrap the callback to be used imperatively. It would probably be much better to put the rest of the function in the callback instead, this current way is admittedly setting off my bug senses.

I mostly want to check that the public API concept is sound, and if so let's tidy this up.

Reply:
> I feel like a `connect`-style-middleware would be good

I'm not sure what pattern exactly this is referring to, could you point me in the right direction? Definitely happy to explore alternative approaches :)

Reply:
`req,res,next`-thing https://github.com/senchalabs/connect

Reply:
Got it, so on the public-api front, I think connect would let people compose together a multi-stage middleware if we just accept one `NodeMiddlewareLike` (maybe rename cors->middleware)

On the internal side I'm having a play with restructuring as a middleware in a 0-dependency way, and I honestly think I'm just making the code more complex 😆, this is working but I think mostly just obfuscates the code:

```ts
/* eslint-disable @typescript-eslint/no-non-null-assertion */
import { AnyRouter } from '../../core';
import { inferRouterContext } from '../../core/types';
import { HTTPRequest } from '../../http/internals/types';
import { resolveHTTPResponse } from '../../http/resolveHTTPResponse';
import { getPostBody } from './internals/getPostBody';
import {
  NodeHTTPHandlerOptions,
  NodeHTTPRequest,
  NodeHTTPResponse,
} from './types';

type NodeHTTPRequestHandlerOptions<
  TRouter extends AnyRouter,
  TRequest extends NodeHTTPRequest,
  TResponse extends NodeHTTPResponse,
> = {
  req: TRequest;
  res: TResponse;
  path: string;
} & NodeHTTPHandlerOptions<TRouter, TRequest, TResponse>;

interface RequestHandler<
  TRouter extends AnyRouter,
  TRequest extends NodeHTTPRequest,
  TResponse extends NodeHTTPResponse,
> {
  (
    opts: NodeHTTPRequestHandlerOptions<TRouter, TRequest, TResponse>,
    next: (err?: any) => void,
  ): void | Promise<void>;
}

export async function nodeHTTPRequestHandler<
  TRouter extends AnyRouter,
  TRequest extends NodeHTTPRequest,
  TResponse extends NodeHTTPResponse,
>(opts: NodeHTTPRequestHandlerOptions<TRouter, TRequest, TResponse>) {
  const middlewares: RequestHandler<TRouter, TRequest, TResponse>[] = [
    corsHandler,
    trpcHandler,
  ].filter(Boolean);

  type Pipeline = (
    opts: NodeHTTPRequestHandlerOptions<TRouter, TRequest, TResponse>,
  ) => Promise<void>;

  const pipeline = middlewares.reduceRight<Pipeline>(
    (next, middleware) => {
      return async (opts) => {
        await middleware(opts, (err) => {
          if (err) {
            throw err;
          }

          return next(opts);
        });
      };
    },
    async () => {
      //
    },
  );

  return await pipeline(opts);
}

async function corsHandler<
  TRouter extends AnyRouter,
  TRequest extends NodeHTTPRequest,
  TResponse extends NodeHTTPResponse,
>(
  opts: NodeHTTPRequestHandlerOptions<TRouter, TRequest, TResponse>,
  next: () => void,
) {
  if (!opts.cors) {
    return next();
  }

  return opts.cors(opts.req, opts.res, (err) => {
    if (err) {
      throw err;
    }

    next();
  });
}

async function trpcHandler<
  TRouter extends AnyRouter,
  TRequest extends NodeHTTPRequest,
  TResponse extends NodeHTTPResponse,
>(
  opts: NodeHTTPRequestHandlerOptions<TRouter, TRequest, TResponse>,
  next: () => void,
) {
  const createContext = async function _createContext(): Promise<
    inferRouterContext<TRouter>
  > {
    return await opts.createContext?.(opts);
  };
  const { path, router } = opts;

  const bodyResult = await getPostBody(opts);

  const query = opts.req.query
    ? new URLSearchParams(opts.req.query as any)
    : new URLSearchParams(opts.req.url!.split('?')[1]);
  const req: HTTPRequest = {
    method: opts.req.method!,
    headers: opts.req.headers,
    query,
    body: bodyResult.ok ? bodyResult.data : undefined,
  };
  const result = await resolveHTTPResponse({
    batching: opts.batching,
    responseMeta: opts.responseMeta,
    path,
    createContext,
    router,
    req,
    error: bodyResult.ok ? null : bodyResult.error,
    onError(o) {
      opts?.onError?.({
        ...o,
        req: opts.req,
      });
    },
  });

  const { res } = opts;
  if ('status' in result && (!res.statusCode || res.statusCode === 200)) {
    res.statusCode = result.status;
  }
  for (const [key, value] of Object.entries(result.headers ?? {})) {
    if (typeof value === 'undefined') {
      continue;
    }
    res.setHeader(key, value);
  }

  res.end(result.body);

  return next();
}

```

Reply:
This second approach looks much tidier and isn't trying to be too clever:

```ts
/* eslint-disable @typescript-eslint/no-non-null-assertion */
import { AnyRouter } from '../../core';
import { inferRouterContext } from '../../core/types';
import { HTTPRequest } from '../../http/internals/types';
import { resolveHTTPResponse } from '../../http/resolveHTTPResponse';
import { getPostBody } from './internals/getPostBody';
import {
  NodeHTTPHandlerOptions,
  NodeHTTPRequest,
  NodeHTTPResponse,
} from './types';

type NodeHTTPRequestHandlerOptions<
  TRouter extends AnyRouter,
  TRequest extends NodeHTTPRequest,
  TResponse extends NodeHTTPResponse,
> = {
  req: TRequest;
  res: TResponse;
  path: string;
} & NodeHTTPHandlerOptions<TRouter, TRequest, TResponse>;

export async function nodeHTTPRequestHandler<
  TRouter extends AnyRouter,
  TRequest extends NodeHTTPRequest,
  TResponse extends NodeHTTPResponse,
>(opts: NodeHTTPRequestHandlerOptions<TRouter, TRequest, TResponse>) {
  const handleViaMiddleware =
    opts.cors ??
    ((_req, _res, next) => {
      next();
    });

  return handleViaMiddleware(opts.req, opts.res, async () => {
    const createContext = async function _createContext(): Promise<
      inferRouterContext<TRouter>
    > {
      return await opts.createContext?.(opts);
    };
    const { path, router } = opts;

    const bodyResult = await getPostBody(opts);

    const query = opts.req.query
      ? new URLSearchParams(opts.req.query as any)
      : new URLSearchParams(opts.req.url!.split('?')[1]);
    const req: HTTPRequest = {
      method: opts.req.method!,
      headers: opts.req.headers,
      query,
      body: bodyResult.ok ? bodyResult.data : undefined,
    };
    const result = await resolveHTTPResponse({
      batching: opts.batching,
      responseMeta: opts.responseMeta,
      path,
      createContext,
      router,
      req,
      error: bodyResult.ok ? null : bodyResult.error,
      onError(o) {
        opts?.onError?.({
          ...o,
          req: opts.req,
        });
      },
    });

    const { res } = opts;
    if ('status' in result && (!res.statusCode || res.statusCode === 200)) {
      res.statusCode = result.status;
    }
    for (const [key, value] of Object.entries(result.headers ?? {})) {
      if (typeof value === 'undefined') {
        continue;
      }
      res.setHeader(key, value);
    }
    res.end(result.body);
  });
}

```